elastic/logstash
# Json normalization performance

## Release notes

 - Restores performance of JSON-encoding operation by deferring an encoding operation and reducing unnecessary memory copying


## What does this PR do?

 - Eliminates wasteful computation when encoding JSON

## Why is it important/What is the impact to the user?

The safety added by 8.14.1's unicode normalization came at the cost of requiring all deeply-nested strings to be encoded to utf-8 prior to being passed to jrjackson, but separate upstream fixes in jrjackson eliminated this need for _most_ cases.

This PR restores some of the lost performance while keeping identical behaviour. We can chase down the remaining performance gap (which may require a behavior change) in a separate effort.

## Checklist


## How to test this PR locally

1. export an environment variable containing deeply-nested JSON structure
   ~~~
   export MESSAGE='{"one":{"two":{"three":{"four":{"five":{"six":{"seven":{"eight":{"nine":{"ten":{"eleven":{"twelve":{"thirteen":{"fourteen":{"fifteen":{}}}}}}}}}}}}}}}}'
   ~~~

2. use that environment variable as input to a simple logstash pipeline with generator and json codec that invokes `LogStash::Json.dump`:
   ~~~
   bin/logstash -e 'input { generator { threads => 4 message => "${MESSAGE}" codec => json } } filter { ruby { code => "::LogStash::Json.dump(event.to_hash)" } } output { sink { } }'
   ~~~

3. observe throughput:
   ~~~
   watch --color '(curl --silent -XGET http://localhost:9600/_node/stats | jq --color-output "pick(.process, .flow)")'
   ~~~

   On my M3 laptop, this translated to:
   | scenario | valid utf-8 | preserved utf-8 | throughput | cpu | 
   |----------|-------------|-----------------|------------|-----|
   | no normalization (8.13.x)   | ⚠️           | ✅               | 440k eps     | 48%    |
   | pre-normalization (8.14.2)   | ✅           | ✅              | 270k eps   | 62% |
   | jackson normalization (8.14.2)   | ✅           | ⚠️              | 440k eps   | 49% |
   | filtered pre-normalization ([POC](https://gist.github.com/yaauie/c71541ae5002e69f2f0aaff8d6b20677))  | ✅           | ✅              | 320k eps   | 57% |
   | reduced copy pre-normalization (this PR)  | ✅           | ✅              | **360k eps**   | **55%** |
   
   > - `valid utf-8`:
   >   - ✅ : output is _always_ valid `UTF-8`, a pre-requisite for being valid JSON
   >   - ⚠️ : when inputs are non-`UTF-8` or invalid-`UTF-8`, output can be corrupt
   > - `preserved utf-8`:
   >   - ✅ : valid `UTF-8` sequences in binary-flagged strings are preserved
   >   - ⚠️ : each byte in a valid multibyte `UTF-8` sequences in binary-flagged strings is replaced with the unicode replacement character (lossy)
